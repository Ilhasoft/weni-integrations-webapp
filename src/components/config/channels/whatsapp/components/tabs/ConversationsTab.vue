<template>
  <div class="conversations">
    <div class="conversations__content">
      <div :class="['conversations__content__dropdown', { active: showDateFilter }]">
        <div class="conversations__content__label">
          {{ $t('WhatsApp.config.conversations.filter.label') }}
        </div>
        <unnnic-date-filter
          class="conversations__content__dropdown-filter"
          dateFormat="DD/MM/YYYY"
          @filter="showDateFilter = true"
          :startDate="startDateObject"
          :endDate="endDateObject"
          placeholder="DD/MM/YYYY ~ DD/MM/YYYY"
        />

        <div class="conversations__content__dropdown-data">
          <unnnic-date-picker
            v-if="showDateFilter"
            size="small"
            :days="['D', 'S', 'T', 'Q', 'Q', 'S', 'S']"
            @change="handleDateFilter"
          />
        </div>
      </div>

      <div>
        <div class="conversations__content__label">
          {{ $t('WhatsApp.config.conversations.conversations_count.label') }}
        </div>

        <conversations-table
          :userMessages="userInitiated"
          :businessMessages="businessInitiated"
          :total="totalInitiated"
        />
      </div>

      <div class="conversations__content__report">
        <span class="conversations__content__report__label">
          {{ $t('WhatsApp.config.conversations.report.label') }}
        </span>

        <span class="conversations__content__report__description">
          {{ $t('WhatsApp.config.conversations.report.description') }}
        </span>

        <unnnic-tool-tip
          class="conversations__content__report__tooltip"
          :text="$t('WhatsApp.config.conversations.report.missing_dates')"
          :enabled="!this.startDate || !this.endDate"
          side="top"
          maxWidth="15rem"
        >
          <unnnic-button
            class="conversations__content__report__button"
            type="secondary"
            size="small"
            @click="requestReport"
            :disabled="!this.startDate || !this.endDate"
          >
            {{ $t('WhatsApp.config.conversations.report.button') }}
          </unnnic-button>
        </unnnic-tool-tip>
      </div>
    </div>
    <div class="conversations__buttons">
      <unnnic-button
        class="conversations__buttons__close"
        type="secondary"
        size="large"
        :text="$t('general.Close')"
        @click="() => this.$emit('close')"
      />
    </div>
  </div>
</template>

<script>
  import debounce from 'lodash.debounce';
  import ConversationsTable from '../ConversationsTable.vue';
  import { unnnicCallAlert } from '@weni/unnnic-system';
  import { mapActions, mapState } from 'vuex';
  import removeEmpty from '../../../../../../utils/clean';

  export default {
    name: 'ConversationsTab',
    components: {
      ConversationsTable,
    },
    props: {
      app: {
        type: Object,
        default: /* istanbul ignore next */ () => {},
      },
    },
    data() {
      return {
        showDateFilter: false,
        startDate: null,
        endDate: null,
        businessInitiated: 0,
        userInitiated: 0,
        totalInitiated: 0,
      };
    },
    created() {
      /* istanbul ignore next */
      window.addEventListener('click', (event) => {
        if (event.target.closest('.conversations__content__dropdown')) {
          return false;
        }
        this.showDateFilter = false;
      });
    },
    computed: {
      ...mapState({
        project: (state) => state.auth.project,
      }),
      ...mapState('WhatsApp', [
        'whatsAppConversations',
        'loadingConversations',
        'errorConversations',
        'loadingConversationsReport',
        'errorConversationsReport',
      ]),
      startDateObject() {
        return this.startDate && new Date(this.startDate.replace('-', ' '));
      },
      endDateObject() {
        return this.endDate && new Date(this.endDate.replace('-', ' '));
      },
    },
    methods: {
      ...mapActions('WhatsApp', ['getConversations', 'requestConversationsReport']),
      handleDateFilter: debounce(async function (event) {
        this.startDate = event.startDate;
        this.endDate = event.endDate;

        const params = {
          start: event.startDate,
          end: event.endDate,
        };

        await this.getConversations({
          code: this.app.code,
          appUuid: this.app.uuid,
          params,
        });

        if (this.errorConversations) {
          unnnicCallAlert({
            props: {
              text: this.$t('WhatsApp.config.conversations.fetch_error'),
              title: this.$t('general.error'),
              icon: 'alert-circle-1-1',
              scheme: 'feedback-red',
              position: 'bottom-right',
              closeText: this.$t('general.Close'),
            },
            seconds: 6,
          });
          return;
        }

        this.businessInitiated = this.whatsAppConversations.business_initiated;
        this.userInitiated = this.whatsAppConversations.user_initiated;
        this.totalInitiated = this.whatsAppConversations.total;
      }, 750),
      async requestReport() {
        const params = removeEmpty({
          project_uuid: this.project,
          start_date: this.startDate,
          end_date: this.endDate,
        });

        await this.requestConversationsReport({
          code: this.app.code,
          appUuid: this.app.uuid,
          params,
        });

        if (this.errorConversationsReport) {
          let errorText = this.$t('WhatsApp.config.conversations.report_error');
          let errorColor = 'feedback-red';
          let errorTitle = this.$t('general.error');

          if (this.errorConversationsReport.response.status === 409) {
            errorText = this.$t('WhatsApp.config.conversations.report_already_processing');
            errorColor = 'feedback-yellow';
            errorTitle = this.$t('general.attention');
          }

          unnnicCallAlert({
            props: {
              text: errorText,
              title: errorTitle,
              icon: 'alert-circle-1-1',
              scheme: errorColor,
              position: 'bottom-right',
              closeText: this.$t('general.Close'),
            },
            seconds: 6,
          });
          return;
        }

        unnnicCallAlert({
          props: {
            text: this.$t('WhatsApp.config.conversations.report_success'),
            title: this.$t('general.success'),
            icon: 'check-circle-1-1',
            scheme: 'feedback-green',
            position: 'bottom-right',
            closeText: this.$t('general.Close'),
          },
          seconds: 6,
        });
      },
    },
  };
</script>

<style lang="scss" scoped>
  .conversations {
    display: flex;
    flex-direction: column;
    flex: 1;
    margin-top: $unnnic-spacing-inline-xs;

    &__content {
      display: flex;
      flex-direction: column;
      flex: 1;
      gap: $unnnic-spacing-stack-md;

      &__dropdown {
        width: fit-content;
        &-data {
          position: absolute;
          z-index: 1;
        }

        &-filter {
          width: 14rem;
        }
      }

      &__label {
        font-weight: $unnnic-font-weight-bold;
        font-size: $unnnic-font-size-body-gt;
        line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
        color: $unnnic-color-neutral-darkest;

        margin-bottom: $unnnic-spacing-stack-xs;
      }

      &__report {
        display: flex;
        flex-direction: column;
        gap: $unnnic-spacing-stack-sm;
        font-size: $unnnic-font-size-body-gt;
        line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;

        &__label {
          font-weight: $unnnic-font-weight-bold;
          color: $unnnic-color-neutral-darkest;
        }

        &__description {
          font-weight: $unnnic-font-weight-regular;
          color: $unnnic-color-neutral-cloudy;
        }

        &__tooltip {
          width: 33%;
        }

        &__button {
          width: 100%;
        }
      }
    }

    &__buttons {
      display: flex;
      justify-content: flex-end;
      margin-top: $unnnic-spacing-stack-sm;

      &__close {
        width: 50%;
      }
    }
  }
</style>
