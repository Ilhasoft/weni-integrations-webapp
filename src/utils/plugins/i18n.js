import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
