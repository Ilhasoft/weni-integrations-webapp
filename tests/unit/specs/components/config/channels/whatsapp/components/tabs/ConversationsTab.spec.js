jest.mock('lodash.debounce', () => jest.fn((fn) => fn));
import { unnnicCallAlert as mockUnnnicCallAlert } from '@weni/unnnic-system';

jest.mock('@weni/unnnic-system', () => ({
  ...jest.requireActual('@weni/unnnic-system'),
  unnnicCallAlert: jest.fn(),
}));

import VueRouter from 'vue-router';
import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import ConversationsTab from '@/components/config/channels/whatsapp/components/tabs/ConversationsTab.vue';
import i18n from '@/utils/plugins/i18n';
import { singleApp } from '../../../../../../../../__mocks__/appMock';

const localVue = createLocalVue();
localVue.use(VueRouter);
localVue.use(Vuex);

describe('whatsapp/components/tabs/ConversationsTab.vue', () => {
  let wrapper;

  let actions;
  let store;

  beforeEach(() => {
    actions = {
      getConversations: jest.fn(() => {
        return { data: singleApp };
      }),
    };

    store = new Vuex.Store({
      actions,
    });

    wrapper = shallowMount(ConversationsTab, {
      localVue,
      i18n,
      store,
      propsData: {
        app: singleApp,
      },
      mocks: {
        $t: () => 'some specific text',
      },
      stubs: {
        ConversationsTable: true,
        UnnnicDateFilter: true,
        UnnnicDatePicker: true,
        UnnnicButton: true,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be rendered properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleDateFilter()', () => {
    it('should call getConversations()', async () => {
      const event = {
        startDate: '03-12-22',
        endDate: '04-13-23',
      };
      expect(actions.getConversations).not.toHaveBeenCalled();
      await wrapper.vm.handleDateFilter(event);
      expect(actions.getConversations).toHaveBeenCalledTimes(1);
      expect(actions.getConversations).toBeCalledWith(expect.any(Object), {
        code: singleApp.code,
        appUuid: singleApp.uuid,
        params: {
          start: event.startDate,
          end: event.endDate,
        },
      });
    });

    it('should call unnnicCallAlert on error', async () => {
      actions.getConversations.mockImplementation(() => {
        throw new Error('error fetching');
      });

      const event = {
        startDate: '03-12-22',
        endDate: '04-13-23',
      };
      expect(mockUnnnicCallAlert).not.toHaveBeenCalled();
      await wrapper.vm.handleDateFilter(event);
      expect(mockUnnnicCallAlert).toHaveBeenCalledTimes(1);
    });
  });
});
