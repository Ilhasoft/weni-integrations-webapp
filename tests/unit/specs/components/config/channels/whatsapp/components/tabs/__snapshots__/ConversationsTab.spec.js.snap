// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whatsapp/components/tabs/ConversationsTab.vue should be rendered properly 1`] = `
<div class="conversations">
  <div class="conversations__content">
    <div class="conversations__content__dropdown">
      <div class="conversations__content__label">
        Filter by
      </div>
      <unnnicdatefilter-stub size="sm" dateformat="DD/MM/YYYY" dateseparator="~" placeholder="DD/MM/YYYY ~ DD/MM/YYYY" class="conversations__content__dropdown-filter"></unnnicdatefilter-stub>
      <div class="conversations__content__dropdown-data">
        <!---->
      </div>
    </div>
    <div class="conversations__content__tables">
      <div class="conversations__content__label">
        Number of conversations
      </div>
      <div class="conversations__content__tables__section">
        <!---->
        <!---->
      </div>
      <div class="conversations__content__tables__section">
        <!---->
        <unnnic-table-stub items="[object Object],[object Object],[object Object],[object Object]">
          <unnnic-table-row-stub headers="[object Object],[object Object]"></unnnic-table-row-stub>
          <unnnic-table-row-stub headers="[object Object],[object Object]"></unnnic-table-row-stub>
        </unnnic-table-stub>
      </div>
    </div>
    <div class="conversations__content__report"><span class="conversations__content__report__label">
        Detailed conversation report
      </span> <span class="conversations__content__report__description">
        You can receive message breakdowns by email of no more than 6 months ago. It is possible that some information in the report is inaccurate. If in doubt, contact your account manager.
      </span>
      <unnnic-tool-tip-stub text="A date range is required to be able to request a report" enabled="true" side="top" maxwidth="15rem" class="conversations__content__report__tooltip">
        <unnnicbutton-stub size="small" type="secondary" disabled="true" class="conversations__content__report__button">
          Request report
        </unnnicbutton-stub>
      </unnnic-tool-tip-stub>
    </div>
    <div class="conversations__content__insights">
      <div class="conversations__content__insights__title">
        Insights
      </div>
      <div class="conversations__content__insights__about">
        <p>
          See more detailed data related to sending, delivering and reading triggered message templates.
        </p>
      </div>
      <div class="conversations__content__insights__button">
        <unnnicbutton-stub size="large" text="See insights" type="primary"></unnnicbutton-stub>
      </div>
    </div>
  </div>
  <div class="conversations__buttons">
    <unnnicbutton-stub size="large" text="Cancel" type="secondary" class="conversations__buttons__close"></unnnicbutton-stub>
  </div>
</div>
`;
